### DISCUSSION TITLE
What is the best way to serve static files after authenticating with remix-auth?

### DISCUSSION THREAD

**Discussion by vedantroy** [remix-run/remix#3514] [Q&A]
It's pretty easy to serve static files inside of the public folder, but I'm using [remix-auth](https://github.com/sergiodxa/remix-auth), and I'd like to serve a static file only if the user has been authenticated.

It's somewhat easy to do by reading the entire file into an `ArrayBuffer` and then constructing a response with the `ArrayBuffer`, but ideally, I'd be able to serve the file with support for things like range requests, and the `ArrayBuffer` technique makes this impossible.

**Comment by kiliman** [comment]
What adapter are you using? If you use Express you can check for the auth cookie in your static file handler.

**Reply by vedantroy** [reply]
I was using the default remix app server, although I suppose I could switch to express.

**Reply by kiliman** [reply]
I have a script that will let you eject from RAS to Express.

https://github.com/kiliman/rmx-cli#eject-ras

**Reply by vedantroy** [reply]
The issue is, it's not enough to just check for the auth cookie -- I also want to do database calls based on the user id stored in the auth cookie. So I need to deserialize the auth cookie, (I believe normally remix-auth handles this?)

**Reply by kiliman** [reply]
I'm not sure I understand. I thought you needed to protect static files. Typically you let the server handle that so you don't have to read the file directly. That's why I said to check the cookie in your static file handler.

Are you saying you can't trust the cookie and need to verify it against the database first?

**Reply by sergiodxa** [reply]
Maybe the cookie has a userId and he needs to do a DB query to get the user and check if it has permissions to read the file.

I would do it inside a resource route, then use `fs.readFile` to read the file and send it.

**Reply by vedantroy** [reply]
> Maybe the cookie has a userId and he needs to do a DB query to get the user and check if it has permissions to read the file.
> 
> I would do it inside a resource route, then use `fs.readFile` to read the file and send it.

This is exactly correct. The issue is, doing `fs.readFile` is not very fast because I believe it sends back the file as a giant binary blob. This isn't optimal because there is no support for range requests.

E.g, PDF.js supports using range-requests to load a file in small chunks, which allows users to see a PDF much faster.
With this current method, my loading times for a single file are on the order of 10+ seconds instead of ~ 1 second.

**Reply by sergiodxa** [reply]
It should be possible to read the file as a stream and send the stream to the browser so you send byte by byte ID the file as your read it
